backends/gstreamer/src/source.rs

```diff
@@ -0,0 +1,236 @@
+                unsafe {
+                    TYPE = t;
+                }
+            });
+
+            unsafe { TYPE }
+        }
+
+        fn class_init(klass: &mut BinClass) {
+            klass.set_metadata(
+                "Servo Media Source",
+                "Source/Audio/Video",
+                "Feed player with media data",
+                "Servo developers",
+            );
+
+            let caps = gst::Caps::new_simple(
+                "video/x-raw",
+                &[
+                    ("format", &"BGRA"),
+                    ("pixel-aspect-ratio", &gst::Fraction::from((1, 1))),
+                ],
+            );
+
+            let src_pad_template = gst::PadTemplate::new(
+                "src",
+                gst::PadDirection::Src,
+                gst::PadPresence::Always,
+                &caps,
+            );
+
+            klass.add_pad_template(src_pad_template);
+        }
+
+        pub fn set_size(&self, size: i64) {
+            if self.appsrc.get_size() != -1 {
+                return;
+            }
+            self.appsrc.set_size(size);
+        }
+
+        inner_appsrc_proxy!(end_of_stream, gst::FlowReturn);
+        inner_appsrc_proxy!(get_current_level_bytes, u64);
+        inner_appsrc_proxy!(get_max_bytes, u64);
+        inner_appsrc_proxy!(push_buffer, buffer, gst::Buffer, gst::FlowReturn);
+        inner_appsrc_proxy!(set_callbacks, callbacks, AppSrcCallbacks, ());
+        inner_appsrc_proxy!(set_max_bytes, bytes, u64, ());
+        inner_appsrc_proxy!(set_property_block, block, bool, ());
+        inner_appsrc_proxy!(set_property_format, format, gst::Format, ());
+        inner_appsrc_proxy!(set_stream_type, type_, AppStreamType, ());
+    }
+
+    impl ObjectImpl<Bin> for ServoSrc {
+        fn constructed(&self, bin: &Bin) {
+            bin.parent_constructed();
+
+            self.add_element(bin, &self.appsrc.clone().upcast());
+
+            let pad = self
+                .appsrc
+                .get_static_pad("src")
+                .expect("Could not get src pad");
+
+            let ghost_pad =
+                gst::GhostPad::new("src", &pad).expect("Could not create src ghost pad");
+
+            // In order to make buffering/downloading work as we want, apart from
+            // setting the appropriate flags on the player playbin,
+            // the source needs to either:
+            //
+            // 1. be an http, mms, etc. scheme
+            // 2. report that it is "bandwidth limited".
+            //
+            // 1. is not straightforward because we are using a servosrc scheme for now.
+            // This may change in the future if we end up handling http/https/data
+            // URIs, which is what WebKit does.
+            //
+            // For 2. we need to make servosrc handle the scheduling properties query
+            // to report that it "is bandwidth limited".
+            ghost_pad.add_probe(gst::PadProbeType::QUERY_UPSTREAM, |_, probe_info| {
```

Comment:
I'd recommend overriding the query function of the ghost pad. And for the default handling (for all other queries), you can call `gst::Pad::query_default()`.

Reply:
I applied @ceyusa's [suggestion](https://github.com/servo/media/pull/179#discussion_r247875364) so I think this is no longer required.